agents: make tcp/udp inherit from virtual class

trevnorris opened the pull request:
Make it so StatsDTcp and StatsDUdp inherit from the virtual class StatsDConnection. This way the StatsDAgent only needs to keep a single reference to whichever class is being used. This will also make it easier to have multiple instances of StatsDAgent.

trevnorris committed:
agents: make tcp/udp inherit from virtual class

Make it so StatsDTcp and StatsDUdp inherit from the virtual class
StatsDConnection. This way the StatsDAgent only needs to keep a single
reference to whichever class is being used. This will also make it
easier to have multiple instances of StatsDAgent.

PR-UR: https://github.com/nodesource/nsolid/pull/122
Reviewed-by: Santiago Gimeno <santiago.gimeno@gmail.com>

trevnorris commented on agents/statsd/src/statsd_agent.cc:
```diff
@@ -838,18 +841,17 @@ int StatsDAgent::config_handles() {
 
+  if (connection_) {
+    connection_->close_and_delete();
+    connection_ = nullptr;
```
I suppose this could be improved by checking if `connection_->endpoint()->type() == endpoint->type()`, and reusing it if that's the case.

santigimeno commented on agents/statsd/src/statsd_agent.h:
```diff
@@ -284,9 +315,8 @@ class StatsDAgent: public std::enable_shared_from_this<StatsDAgent> {
 
-  StatsDTcp* tcp_;
+  StatsDConnection* connection_ = nullptr;
   std::unique_ptr<StatsDUdp> udp_;
-  std::unique_ptr<StatsDEndpoint> endpoint_;
   nsuv::ns_timer retry_timer_;
```
you should be able to remove `udp_` and `retry_timer_`

trevnorris replied:
Whoops. Forgot those. Thanks.

santigimeno commented on agents/statsd/src/statsd_endpoint.h:
```diff
@@ -49,6 +49,13 @@ class StatsDEndpoint {
+    Udp,
+  };
+
+  Type type() const { return type_; }
```
```suggestion
  constexpr Type type() const { return type_; }
```

trevnorris replied:
will do

santigimeno reviewed (approved):
LGTM with a couple of nits

trevnorris committed:
tools: add agents path to workflow

PR-UR: https://github.com/nodesource/nsolid/pull/122
Reviewed-by: Santiago Gimeno <santiago.gimeno@gmail.com>

trevnorris committed:
agents: set state when writing

Need to let the instance know when it's attempting to write so it
doesn't delete itself in the middle of writing out data.

PR-UR: https://github.com/nodesource/nsolid/pull/122
Reviewed-by: Santiago Gimeno <santiago.gimeno@gmail.com>
